stacks-signer/src/v0/signer_state.rs

```diff
@@ -38,10 +40,234 @@ use crate::signerdb::SignerDb;
+                active_signer_protocol_version,
+            };
+            let entry = state_views
+                .entry(state_machine.clone())
+                .or_insert_with(|| 0);
+            *entry += weight;
+            if self.reached_agreement(*entry) {
+                return Some(state_machine);
+            }
+        }
+        None
+    }
+
+    /// Check if there is an agreed upon global current miner viewpoint
+    pub fn determine_global_current_miner(
+        &mut self,
+        local_address: StacksAddress,
+        local_update: StateMachineUpdateMessage,
+    ) -> Option<StateMachineUpdateMinerState> {
+        let (global_burn_block, _) =
+            self.determine_global_burn_view(local_address, local_update)?;
+        let mut miner_views = HashMap::new();
+        for (address, update) in &self.address_updates {
+            let Some(weight) = self.address_weights.get(address) else {
+                continue;
+            };
+            let StateMachineUpdateContent::V0 {
+                burn_block,
+                current_miner,
+                ..
+            } = &update.content;
+
+            if *burn_block != global_burn_block {
+                continue;
+            }
+
+            let entry = miner_views.entry(current_miner).or_insert_with(|| 0);
+            *entry += weight;
+            if self.reached_agreement(*entry) {
+                return Some(current_miner.clone());
+            }
+        }
+        None
+    }
+
+    /// Determines whether a signer with the `local_address` and `local_update` should
+    /// should capitulate its current miner view to a new state. This is not necessarily the same as the current global view
+    /// of the miner as it is up to signers to capitulate before this becomes the finalized view.
+    pub fn capitulate_miner_view(
+        &mut self,
+        signerdb: &mut SignerDb,
+        local_address: StacksAddress,
+        local_update: StateMachineUpdateMessage,
+    ) -> Option<StateMachineUpdateMinerState> {
+        let StateMachineUpdateContent::V0 {
+            burn_block: current_burn_block,
+            ..
+        } = local_update.content.clone();
+        let (global_burn_view, _) = self.determine_global_burn_view(local_address, local_update)?;
+        if current_burn_block != global_burn_view {
+            return None;
+        }
+        let mut current_miners = HashMap::new();
+        for (address, update) in &self.address_updates {
+            let Some(weight) = self.address_weights.get(address) else {
+                continue;
+            };
+            let StateMachineUpdateContent::V0 {
+                burn_block,
+                current_miner,
+                ..
+            } = &update.content;
+
+            if *burn_block != global_burn_view {
+                continue;
+            }
+
+            let StateMachineUpdateMinerState::ActiveMiner { tenure_id, .. } = current_miner.clone()
```

Comment:
Similar here, this clone is not necessary. I think for this one you can just use a reference then remove the `&` on line 237.